subsys/bluetooth/host/hci_core.c

```diff
@@ -7123,6 +7291,339 @@ bool bt_addr_le_is_bonded(uint8_t id, const bt_addr_le_t *addr)
+	}
+
+	if (param->options & LE_PER_SYNC_OPT_DONT_SYNC_AOD_1US) {
+		cp->cte |= BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_NO_AOD_1US;
+	}
+
+	if (param->options & LE_PER_SYNC_OPT_DONT_SYNC_AOD_2US) {
+		cp->cte |= BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_NO_AOD_2US;
+	}
+
+	if (param->options & LE_PER_SYNC_OPT_SYNC_ONLY_CONST_TONE_EXT) {
+		cp->cte |= BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_ONLY_CTE;
+	}
+
+	cp->sid = param->sid;
+	cp->skip = param->skip;
+	cp->sync_timeout = param->timeout;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_PER_ADV_CREATE_SYNC, buf, NULL);
+	if (err) {
+		per_adv_sync_delete(per_adv_sync);
+		return err;
+	}
+
+	syncing_per_adv_sync = per_adv_sync;
+
+	/* Syncing requires that scan is enabled. If the callee doesn't enable
+	 * scan first, we enable it here, and disable it once the sync has been
+	 * established. We don't need to use any callbacks since we rely on
+	 * the advertiser address in the sync params.
+	 */
+	if (!atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING)) {
+		err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, NULL);
+		atomic_set_bit(per_adv_sync->flags,
+			       BT_PER_ADV_SYNC_SCAN_INITIATED);
+
+		if (err) {
+			bt_le_per_adv_sync_delete(per_adv_sync);
+			return err;
+		}
+	}
+
+	*out_sync = per_adv_sync;
+	memcpy(&per_adv_sync->params, param, sizeof(*param));
+	per_adv_sync->cb = cb;
+
+	return 0;
+}
+
+static int bt_le_per_adv_sync_create_cancel(
+	struct bt_le_per_adv_sync *per_adv_sync)
+{
+	struct net_buf *buf;
+	int err;
+
+	if (syncing_per_adv_sync != per_adv_sync) {
+		return -EINVAL;
+	}
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_PER_ADV_CREATE_SYNC_CANCEL, 0);
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_PER_ADV_CREATE_SYNC_CANCEL, buf,
+				   NULL);
+	if (err) {
+		return err;
+	}
+
+	per_adv_sync_delete(per_adv_sync);
+
+	syncing_per_adv_sync = NULL;
+
+	return 0;
+}
+
+static int bt_le_per_adv_sync_terminate(struct bt_le_per_adv_sync *per_adv_sync)
+{
+	struct bt_hci_cp_le_per_adv_terminate_sync *cp;
+	struct net_buf *buf;
+	int err;
+
+	if (!atomic_test_bit(per_adv_sync->flags, BT_PER_ADV_SYNC_SYNCED)) {
+		return -EINVAL;
+	}
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_PER_ADV_TERMINATE_SYNC,
+				sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = per_adv_sync->handle;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_PER_ADV_TERMINATE_SYNC, buf,
+				   NULL);
+	if (err) {
+		return err;
+	}
+
+	per_adv_sync_delete(per_adv_sync);
+
+	return 0;
+}
+
+int bt_le_per_adv_sync_delete(struct bt_le_per_adv_sync *per_adv_sync)
+{
+	int err = 0;
+
+	if (atomic_test_bit(per_adv_sync->flags, BT_PER_ADV_SYNC_SYNCED)) {
+		err = bt_le_per_adv_sync_terminate(per_adv_sync);
+	} else if (syncing_per_adv_sync == per_adv_sync) {
+		err = bt_le_per_adv_sync_create_cancel(per_adv_sync);
+	}
+
+	if (err) {
+		return err;
+	}
+
+	per_adv_sync_delete(per_adv_sync);
```

Comment:
bt_le_per_adv_sync_terminate will call per_adv_sync_delete. It could probably be removed from there.
Same for bt_le_per_adv_sync_terminate, they are only called from this place.

Reply:
Nice catch!